src/consensus/pbft_manager.cpp

```diff
@@ -1455,20 +1455,29 @@ std::pair<vec_blk_t, bool> PbftManager::comparePbftBlockScheduleWithDAGblocks_(P
 
 bool PbftManager::pushCertVotedPbftBlockIntoChain_(taraxa::blk_hash_t const &cert_voted_block_hash,
                                                    std::vector<Vote> const &cert_votes_for_round) {
-  if (!checkPbftBlockValid_(cert_voted_block_hash)) {
-    syncPbftChainFromPeers_(invalid_cert_voted_block, cert_voted_block_hash);
-    return false;
-  }
   auto pbft_block = pbft_chain_->getUnverifiedPbftBlock(cert_voted_block_hash);
   if (!pbft_block) {
-    LOG(log_er_) << "Can not find the cert vote block hash " << cert_voted_block_hash << " in pbft queue";
+    pbft_block = db_->getPbftCertVotedBlock(cert_voted_block_hash);
+    if (!pbft_block) {
+      LOG(log_nf_) << "Can not find the cert voted block hash " << cert_voted_block_hash
+                   << " in both pbft queue and DB";
+      return false;
+    }
+    // Read from DB pushing into unverified queue
+    pbft_chain_->pushUnverifiedPbftBlock(pbft_block);
+  }
+
+  if (!checkPbftBlockValid_(cert_voted_block_hash)) {
```

Comment:
Let's delete this function and use `pbft_chain_->checkPbftBlockValidation()` instead. The call it on 4 places, but on 3 places you basically already have the shared_ptr to block. On the 4th place you can call `block = pbft_chain_->getUnverifiedPbftBlock()` and then `pbft_chain_->checkPbftBlockValidation(block)`

Reply:
I prefer to keep the checkPbftBlockValid_ function. And in the other 3 places have changed to use pbft_chain_->checkPbftBlockValidation(block)

Reply:
Separate function is created usually 1 of these reasons:

- some functionality is used repeatedly so you put it into the function -> not this case
- big chunks of code are logically grouped in functions so the code is cleaner -> not this case
- interface/data encapsulation (getters/setters/need to use locks, etc...) -> not this case

Reply:
You are repeating this code everywhere in fact:
`auto pbft_block = pbft_chain_->getUnverifiedPbftBlock(previous_round_next_voted_value_);
  if (!pbft_block) {
    pbft_block = db_->getPbftCertVotedBlock(previous_round_next_voted_value_);
    if (pbft_block) {
      // PBFT unverified queue empty after node reboot, read from DB pushing back in unverified queue
      pbft_chain_->pushUnverifiedPbftBlock(pbft_block);
    }
  }`

so if you return bool flag you can reuse it and also print log errors based on it. 

And please use it everywhere if possible and delete checkPbftBlockValid_. Or if you want to keep it pls provide some valid reasoj for it

Reply:
I agree with you. Fixed